src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts

```diff
@@ -0,0 +1,212 @@
+// class to recognise if sketchapp shape is a rectangle, a line, a circle(TODO) or something else
+
+class Point {
+    x: number;
+    y: number;
+
+    constructor(strOrX: string | number, y: number = null) {
+        // a string containing the coords is input
+        if (y === null) {
+            const coords: any = this.stringToCoords(strOrX as string);
+            if (coords) {
+                this.x = this.decRound(coords.x);
+                this.y = this.decRound(coords.y);
+            }
+            // the coords numbers are input
+        } else {
+            this.x = this.decRound(strOrX as number);
+            this.y = this.decRound(y);
+        }
+    }
+
+    // 2 decimals rounding
+    decRound(number: number): number {
+        return Math.round(number * 100 + 0.001) / 100;
+    }
+
+    distanceRounded(point: Point): number {
+        return this.decRound(this.distance(point));
+    }
+    distance(point: Point): number {
+        return Math.sqrt(this.distanceSquared(point));
+    }
+    distanceSquaredRounded(point: Point): number {
+        return this.decRound(this.distanceSquared(point));
+    }
+    distanceSquared(point: Point): number {
+        return Math.pow(this.x - point.x, 2) + Math.pow(this.y - point.y, 2);
+    }
+
+    stringToCoords(string: string): RegExpMatchArray['groups'] | boolean {
+        const regex = /{(?<x>\-?\d(?:\.\d+)?(?:e\-?\d+)?),\s?(?<y>\-?\d(?:\.\d+)?(?:e\-?\d+)?)}/;
+        const match = string.match(regex);
+        if (!!match) {
+            return match.groups;
+        }
+        return false;
+    }
+}
+
+class Cluster {
+    points: Point[] = [];
+    barycenter: Point;
+
+    constructor(point: Point = null) {
+        if (point !== null) {
+            this.addPoint(point);
+        }
+    }
+
+    updateBarycenter(): void {
+        let avgX = 0;
+        let avgY = 0;
+        this.points.forEach(point => {
+            avgX += point.x;
+            avgY += point.y;
+        });
+        avgX /= this.points.length;
+        avgY /= this.points.length;
+        this.barycenter = new Point(avgX, avgY);
+    }
+
+    addPoint(point: Point): void {
+        this.points.push(point);
+        this.updateBarycenter();
+    }
+}
+
+export class Shape {
+    Points: Point[];
+    topLeft: Point;
+    topRight: Point;
+    bottomLeft: Point;
+    bottomRight: Point;
+
+    static get error(): number {
```

Comment:
This is the degree of error accepted by the shape recognition service. It is necessary in most case because : 
- shapes are not perfect rectangles/circles when exported from sketchApp
- the shape recognition is quite basic

I will change it to `static readonly error = 0.05;` for more clarity

Reply:
`shapeRecognitionPrecision` or `recognitionPrecision` can be a more explicit name. Otherwise, it can be confused with an error related to a kind of exception 😉